### PR TITLE
feat(expressions): add router-playground-modal [KM-299]

### PR DESCRIPTION
**Pull request by 2eha0**
# Summary
- migrate `<RouterPlaygroundModal />` from kong-admin to public-ui
- bump it in `<RouterForm />` and add some tests
- [RouterPlayground.cy.ts](https://github.com/Kong/public-ui-components/pull/1497/files#diff-43949cf48c9e835e680939008cb0e3d6c6ff313bc79dd3cbfed1e6fda2645752) is skipped for temporary, it requires `wasm` and `Monaco` vite plugins during the component tests running, which we should discuss with core UI team

### CODE REVIEW

File: packages/core/expressions/src/components/RequestCard.vue [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+        appearance="info"
+        class="protocol"
+      >
+        {{ protocol }}
+      </KBadge>
+      <KBadge
+        v-if="['http', 'https'].includes(protocol) && method"
+        :appearance="getMethodBadgeAppearance(method)"
+      >
+        {{ method?.toUpperCase() }}
```
Comment (Leopoldthecoder) [inline packages/core/expressions/src/components/RequestCard.vue:19]: In the above `v-if` we have guaranteed `method` exists:
```html
v-if="['http', 'https'].includes(protocol) && method"
```

So here we don't need `?`:
```suggestion
        {{ method.toUpperCase() }}
```

File: packages/core/expressions/src/components/MonacoEditor.vue [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+  editor = monaco.editor.create(editorRoot.value as HTMLElement, options)
+
+  editor.onDidChangeModelContent((event) => {
+    const value = editor!.getValue()
+    if (props.modelValue !== value) {
+      emits('update:modelValue', value, event)
+    }
+  })
+
+  // emits('editorDidMount', editor)
```
Comment (Leopoldthecoder) [inline packages/core/expressions/src/components/MonacoEditor.vue:94]: Should we remove this line?

File: packages/core/expressions/src/components/RequestCard.vue [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+  display: flex;
+  flex-direction: column;
+  gap: $kui-space-40;
+  justify-content: flex-start;
+  min-width: 0;
+  padding: $kui-space-50;
+  position: relative;
+
+  &::before {
+    border: 1px solid $kui-color-border-neutral-weaker;
```
Comment (Leopoldthecoder) [inline packages/core/expressions/src/components/RequestCard.vue:126]: Can we use design token here?

File: packages/core/expressions/src/components/RequestImportModal.vue [hunk 4]
```diff
@@ -0,0 +1,155 @@
+<template>
+  <KModal
+    action-button-text="Import"
+    class="import-requests-modal"
+    text-align="left"
+    title="Import requests from JSON"
```
Comment (Leopoldthecoder) [inline packages/core/expressions/src/components/RequestImportModal.vue:6]: Could we add an `i18n` entry for this?

File: packages/core/expressions/src/components/RequestImportModal.vue [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    emit('dismiss', true)
+  }
+}
+
+</script>
+
+<style lang="scss" scoped>
+.import-requests-modal {
+  .import-requests-alert {
+    margin-bottom: 16px;
```
Comment (sumimakito) [inline packages/core/expressions/src/components/RequestImportModal.vue:136]: nit: Could we replace the attribute values with design tokens?

File: packages/core/expressions/src/components/RequestModal.vue [hunk 6]
```diff
@@ -0,0 +1,203 @@
+<template>
+  <KModal
+    action-button-text="Add"
+    text-align="left"
+    title="Add a request"
+    :visible="props.visible"
+    @cancel="emit('dismiss', false)"
+    @proceed="handleProceed"
+  >
+    <template #default>
```
Comment (sumimakito) [inline packages/core/expressions/src/components/RequestModal.vue:10]: Would it be better for us to remove this default template?

File: packages/core/expressions/src/components/RequestCard.vue (hunk 3 above)
Comment (2eha0) [reply packages/core/expressions/src/components/RequestCard.vue]: I checked all styles in this commit, and used design tokens as more as possible.

File: packages/core/expressions/src/components/RequestImportModal.vue (hunk 4 above)
Comment (2eha0) [reply packages/core/expressions/src/components/RequestImportModal.vue]: Fixed

File: packages/entities/entities-routes/src/locales/en.json [hunk 7]
```diff
@@ -248,6 +248,9 @@
     "expressions_editor": {
       "loading": "Loading the Expressions editor…",
       "error": "Error occurred while loading the Expressions editor. Please view the console for more details."
+    },
+    "expressionPlayground": {
+      "testLink": "Test with Router Playground"
```
Comment (sumimakito) [inline packages/entities/entities-routes/src/locales/en.json:253]:
```suggestion
      "test_link": "Test with Router Playground"
```

nit: Could we keep the `snake_case` in this file?

File: packages/entities/entities-routes/vite.config.ts [hunk 8]
```diff
@@ -20,6 +20,7 @@ const config = mergeConfig(sharedViteConfig, defineConfig({
     rollupOptions: {
       external: [
         '@kong-ui-public/expressions', // This is optional if we do not use Expressions features
+        '@kong-ui-public/expressions/dist/style.css', // This is optional if we do not use Expressions features
```
Comment (sumimakito) [inline packages/entities/entities-routes/vite.config.ts:23]: Does this mean we need to import this style file on the host app side?

File: packages/entities/entities-routes/src/composables/useToaster.ts [hunk 9]
```diff
@@ -0,0 +1,12 @@
+import { onBeforeUnmount } from 'vue'
+import { ToastManager } from '@kong/kongponents'
+
+export default function useToaster() {
+  const toaster = new ToastManager()
```
Comment (sumimakito) [inline packages/entities/entities-routes/src/composables/useToaster.ts:5]: I noticed we previously removed the initialization of `ToastManager` from this repo in https://github.com/Kong/public-ui-components/issues/932, and decided to let the host apps manage the `ToastManager` instances.

What about passing the toast message as a string in an emitted event to the host apps? Below is an example from `@kong-ui-public/entities-routes`:

https://github.com/Kong/public-ui-components/blob/9349ae1cb9d67be2f0bf050c858346a4f383f489/packages/entities/entities-routes/src/components/RouteList.vue#L229

**Comment by Leopoldthecoder** [comment]
nit: we don't need to include `@kong-ui-public/` in the scope section of the commit message. I'd suggest changing the commit message to `feat(expressions): add router-playground-modal`.

File: .github/workflows/ci.yaml [hunk 10]
```diff
@@ -148,7 +150,17 @@ jobs:
...
                 componentTestFilter="${componentTestFilter}${pkgFolder},"
-                componentTestMatrix="${componentTestMatrix}\"${pkgFolder}\","
+
+                # if there is cypress.config in the package, that means it should run tests independently.
+                # we use `pnpm run test:component:ci --project <pkgFolder>` instead.
+                # so we will not add it to the componentTestMatrix, rather we will add it to a matrix.include
+                # that we can detect it later
+                findCyConfig=$(find "${pkgFolder}" -name "cypress.config.ts" || true)
+                if [[ -z "${findCyConfig}" ]]; then
+                    componentTestMatrix="${componentTestMatrix}\"${pkgFolder}\","
```
Comment (nekolab) [inline .github/workflows/ci.yaml:160]: We should always fill the `componentTestMatrix` and add conditional flag into the `componentTestMatrixInclude`

File: .github/workflows/ci.yaml [hunk 11]
```diff
@@ -207,6 +226,9 @@ jobs:
           # this is component tests in the format json.parse understands
           echo "componentTestMatrix=${componentTestMatrix}" >> $GITHUB_OUTPUT
 
+          # this is component tests in the format json.parse understands
```
Comment (nekolab) [inline .github/workflows/ci.yaml:229]:
```
# this is component tests in the format json.parse understands
echo "componentTestMatrix=${componentTestMatrix}" >> $GITHUB_OUTPUT
echo "runInPackageTestMatrix=${runInPackageTestMatrix}" >> $GITHUB_OUTPUT
```

File: .gitignore [hunk 12]
```diff
@@ -16,9 +16,9 @@ coverage
 _public
 bundle-analyzer
 
-cypress/videos/
-cypress/screenshots/
-cypress/downloads/
+**/cypress/videos/
```
Comment (nekolab) [inline .gitignore:19]: Correct me if I'm wrong. I just remembered that changing `cypress/videos/` to `**/cypress/videos/` seems identical to `git-ignore`

**Review by nekolab — COMMENTED** [review]
Leave some workflow and git comments

File: .gitignore (hunk 12 above)
Comment (nekolab) [reply .gitignore:19]: My bad 🥹

File: packages/core/expressions/src/components/RequestImportModal.vue [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+const props = defineProps<{
+  visible: boolean
+}>()
+
+const emit = defineEmits<{
+  dismiss: [completed: boolean]
+  import: [requests: Request[]]
+}>()
+
+// const playgroundStore = useRouterPlaygroundStore()
```
Comment (Leopoldthecoder) [inline packages/core/expressions/src/components/RequestImportModal.vue:66]: Should we remove this line?

File: packages/core/expressions/src/components/ExpressionsEditor.vue [hunk 14]
```diff
@@ -26,8 +27,10 @@ const props = withDefaults(defineProps<{
   schema: Schema,
   parseDebounce?: number,
   inactiveUntilFocused?: boolean,
+  testId?: string,
 }>(), {
   parseDebounce: 500,
+  testId: 'expressions-editor',
```
Comment (Leopoldthecoder) [inline packages/core/expressions/src/components/ExpressionsEditor.vue:33]: We can remove this prop

File: packages/core/expressions/src/components/PageHeader.vue [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+
+<script setup lang="ts">
+import { defineProps } from 'vue'
+
+defineProps<{
+  title: string,
+  size: number,
+  hideTitle?: boolean,
+  noMargin?: boolean,
+  testId?: string,
```
Comment (Leopoldthecoder) [inline packages/core/expressions/src/components/PageHeader.vue:33]: We can remove this prop

**Review by adamdehaven — CHANGES_REQUESTED** [review]
On first look, I’m not sure I understand why the repository testing flow needs to be altered for this specific package.

File: packages/entities/entities-routes/vite.config.ts (hunk 8 above)
Comment (adamdehaven) [reply packages/entities/entities-routes/vite.config.ts:23]: I’m not sure I understand not bundling the style here. If they are scoped they’ll be excluded.

File: packages/entities/entities-routes/cypress.config.ts [hunk 16]
```diff
@@ -0,0 +1,19 @@
+import { defineConfig } from 'cypress'
```
Comment (adamdehaven) [inline packages/entities/entities-routes/cypress.config.ts:1]: Can you help me understand why you’re adding a completely new cypress component testing setup that replicates the functionality already provided in the repository?

File: packages/core/expressions/package.json [hunk 17]
```diff
@@ -31,12 +31,13 @@
     "stylelint": "stylelint --allow-empty-input './src/**/*.{css,scss,sass,less,styl,vue}'",
     "stylelint:fix": "stylelint --allow-empty-input './src/**/*.{css,scss,sass,less,styl,vue}' --fix",
     "typecheck": "vue-tsc -p './tsconfig.build.json' --noEmit",
-    "test:component": "BABEL_ENV=cypress cross-env FORCE_COLOR=1 cypress run --component -b chrome --spec './src/**/*.cy.ts' --project '../../../.'",
-    "test:component:open": "BABEL_ENV=cypress cross-env FORCE_COLOR=1 cypress open --component -b chrome --project '../../../.'",
+    "test:component": "BABEL_ENV=cypress cross-env FORCE_COLOR=1 cypress run --component -b chrome --spec './src/**/*.cy.ts' --project './'",
```
Comment (adamdehaven) [inline packages/core/expressions/package.json:34]: These changes alter the test runs for this package to be different than the rest of the repository. Help me understand why this change is needed, and why it’s part of this PR?

File: packages/entities/entities-routes/docs/route-form.md [hunk 18]
```diff
@@ -109,7 +109,7 @@ Show/hide Route name field. If `true`, `name` field is stripped from payload obj
 - required: `false`
 - default: `false`
 
-Show/hide Service Select field. Should be used in case of manual adding `service_id` in payload. 
+Show/hide Service Select field. Should be used in case of manual adding `service_id` in payload.
```
Comment (kaiarrowood) [inline packages/entities/entities-routes/docs/route-form.md:112]:
```suggestion
Show/hide Service Select field. Should be used in case of manually adding `service_id` in payload.
```

File: packages/entities/entities-routes/sandbox/pages/RouteFormPage.vue [hunk 19]
```diff
@@ -26,6 +26,7 @@
     :config="konnectConfig"
     :route-flavors="routeFlavors"
     :route-id="routeId"
+    :show-expressions-modal-entry="true"
```
Comment (kaiarrowood) [inline packages/entities/entities-routes/sandbox/pages/RouteFormPage.vue:29]:
```suggestion
    show-expressions-modal-entry
```

nit

File: packages/entities/entities-routes/sandbox/pages/RouteFormPage.vue [hunk 20]
```diff
@@ -55,6 +56,7 @@
     :config="kongManagerConfig"
     :route-flavors="routeFlavors"
     :route-id="routeId"
+    :show-expressions-modal-entry="true"
```
Comment (kaiarrowood) [inline packages/entities/entities-routes/sandbox/pages/RouteFormPage.vue:59]:
```suggestion
    show-expressions-modal-entry
```

nit

File: packages/entities/entities-routes/src/components/RouteFormExpressionsEditor.vue [hunk 21]
```diff
@@ -4,21 +4,108 @@
...
+    placement="bottom-start"
+    :text="tooltipText"
+  >
+    <div
+      v-if="showExpressionsModalEntry"
+      :class="{ disabled: !isHttpBasedProtocol, 'route-form-open-in-playground': true }"
+      data-testid="open-router-playground"
+      @click="handleOpenPlayground"
+    >
+      <RocketIcon size="16px" />
```
Comment (kaiarrowood) [inline packages/entities/entities-routes/src/components/RouteFormExpressionsEditor.vue:18]:
```suggestion
      <RocketIcon size="KUI_ICON_SIZE_30" />
```

suggestion: use from `@kong/design-tokens`

**Review by kaiarrowood — COMMENTED** [review]
Some nits and suggestions. Did not functional test

File: packages/core/expressions/cypress.config.ts [hunk 22]
```diff
@@ -0,0 +1,18 @@
+import { defineConfig } from 'cypress'
+
+export default defineConfig({
```
Comment (adamdehaven) [inline packages/core/expressions/cypress.config.ts:3]: Is this file still needed?

File: packages/entities/entities-routes/vite.config.ts [hunk 23]
```diff
@@ -20,22 +20,28 @@ const config = mergeConfig(sharedViteConfig, defineConfig({
...
+      },
+    }
+    : {}),
+  plugins: [
+    // This plugin is only used in the sandbox & testing environment
+    // It generates extra files in dist folder whitch are not need in library build
+    ...((process.env.USE_SANDBOX || process.env.BABEL_ENV === 'cypress')
+      ? [((monacoEditorPlugin as any).default as typeof monacoEditorPlugin)({})]
+      : []),
+  ],
```
Comment (adamdehaven) [inline packages/entities/entities-routes/vite.config.ts:44]: issue: this plugins block can be removed from this root vite.config since it's in the package version

**Review by ValeryG — CHANGES_REQUESTED** [review]
I think introducing the mix of packages that share cypress.config and another set op packages that requires it's of cypress config Is a mistake. This makes CI (already complicated) even more complicated. 

This makes me as a developer trying to create new package even more confused - do I do it with my own config or do I use shared... 

We need to find the way of doing it unified way. If it has to be cypress config for each package - let's do it for every package. Have shared config as a function and extend it for packages that are needed.

But let's keep CI and package structure universal.

**Comment by adamdehaven** [comment]
I agree with @ValeryG here. We have an existing pattern in `konnect-ui-apps` - if we need to change, we should be more similar across repos.

**Comment by 2eha0** [comment]
@adamdehaven @ValeryG 
Alright, It seems that the current test configuration method needs to be refactored. I'll create another PR to address this, For now, in this PR, I will revert the CI changes and skip some tests.

**Comment by adamdehaven** [comment]
> @adamdehaven @ValeryG Alright, It seems that the current test configuration method needs to be refactored. I'll create another PR to address this, For now, in this PR, I will revert the CI changes and skip some tests.

Rather than refactoring the test configuration, it may be better to write up a summary of why the current setup does not meet your needs, and let the Core UI team evaluate any changes we can implement so that it works better for everyone, rather than integrating changes for a specific package.

File: packages/core/expressions/vite.config.ts [hunk 24]
```diff
@@ -21,17 +21,17 @@ const config = mergeConfig(sharedViteConfig, defineConfig({
       fileName: (format) => `${sanitizedPackageName}.${format}.js`,
     },
     rollupOptions: {
-      external: ['monaco-editor'],
+      external: ['monaco-editor', '@kong-ui-public/forms', '@kong-ui-public/forms/dist/style.css'],
```
[the remark this answers is not in the document]
Comment (adamdehaven) [reply packages/core/expressions/vite.config.ts:24]: This comment was resolved without discussion

[the remark this answers is not in the document]
Comment (2eha0) [reply packages/core/expressions/vite.config.ts:24]: If we only externalize @kong-ui-public/forms, Its styles are still included in the bundle file.

File: packages/entities/entities-routes/vite.config.ts (hunk 8 above)
Comment (adamdehaven) [reply packages/entities/entities-routes/vite.config.ts:23]: > if they do, they should install expressions manually

So how would this be handled in KM or Konnect?

File: packages/core/expressions/vite.config.ts (hunk 24 above)
[the remark this answers is not in the document]
Comment (adamdehaven) [reply packages/core/expressions/vite.config.ts:24]: you should be able to update to a regular expression to cover both, but not a blocker

File: packages/core/expressions/cypress.config.ts (hunk 22 above)
Comment (adamdehaven) [reply packages/core/expressions/cypress.config.ts]: It looks like the file was removed already?

File: packages/entities/entities-routes/cypress.config.ts (hunk 16 above)
Comment (2eha0) [reply packages/entities/entities-routes/cypress.config.ts]: Removed

File: packages/core/expressions/package.json [hunk 25]
```diff
@@ -64,12 +65,15 @@
...
     "@kong/kongponents": "^9.0.21",
     "monaco-editor": "0.21.3",
     "vue": "^3.4.31"
   },
   "dependencies": {
-    "@kong-ui-public/core": "workspace:^"
+    "@kong-ui-public/core": "workspace:^",
+    "@kong-ui-public/i18n": "workspace:^",
+    "@kong/icons": "^1.14.2",
+    "uuid": "^9.0.1"
```
[the remark this answers is not in the document]
Comment (2eha0) [reply packages/core/expressions/package.json:79]: Added

File: packages/core/expressions/cypress.config.ts (hunk 22 above)
Comment (2eha0) [reply packages/core/expressions/cypress.config.ts]: Yes, removed

File: .github/workflows/ci.yaml [hunk 26]
```diff
@@ -27,6 +27,7 @@ jobs:
       unitTestMatrix: ${{steps.changed_packages.outputs.unitTestMatrix}}
       componentTestFilter: ${{steps.changed_packages.outputs.componentTestFilter}}
       componentTestMatrix: ${{steps.changed_packages.outputs.componentTestMatrix}}
+      componentTestMatrixInclude: ${{steps.changed_packages.outputs.componentTestMatrixInclude}}
```
[the remark this answers is not in the document]
Comment (2eha0) [reply .github/workflows/ci.yaml]: Reverted

File: packages/entities/entities-routes/vite.config.ts (hunk 8 above)
Comment (2eha0) [reply packages/entities/entities-routes/vite.config.ts:23]: They installed both of expressions and entities-route
https://github.com/Kong/konnect-ui-apps/pull/3745
https://github.com/Kong/kong-manager/pull/256

**Review by ValeryG — DISMISSED** [review]

**Review by sumimakito — COMMENTED** [review]
Did functionality tests and LGTM. 🚀 

We have some known issues in the router playground:
- HTTP headers in requests appeared as `[object Object]` in the exported JSON
- Support HTTP headers with multiple values (should add context value per header value)
- Better ID handling while importing/exporting requests
- Missing VFG in the sandbox causing Vue warnings

As the main goal of this PR is to move the playground from `kong-admin` to here, we can address them as separate follow-ups to avoid blocking this one.

Please hold off merging this until we have green CI in adoption PRs.

File: packages/core/expressions/README.md [hunk 27]
```diff
@@ -58,9 +60,10 @@ Import the component(s) in your host application as well as the package styles:
 ```ts
 import { asyncInit, ExpressionsEditor } from '@kong-ui-public/expressions'
 import '@kong-ui-public/expressions/dist/style.css'
+import '@kong-ui-public/forms/dist/style.css' // optional: required for `RouterPlaygroundModal` component
```
Comment (sumimakito) [inline packages/core/expressions/README.md:63]: nit: the host app will also need to register the VFG component via code like:

```js
app.component('VueFormGenerator', VueFormGenerator)
```

Otherwise, they would see a warning like:

Comment (2eha0) [reply packages/core/expressions/README.md:63]: fff

Comment (2eha0) [reply packages/core/expressions/README.md:63]: Will be also added in KM-391.

File: packages/entities/entities-routes/src/locales/en.json [hunk 28]
```diff
@@ -248,6 +248,10 @@
     "expressions_editor": {
       "loading": "Loading the Expressions editor…",
       "error": "Error occurred while loading the Expressions editor. Please view the console for more details."
+    },
+    "expressionPlayground": {
```
Comment (sumimakito) [inline packages/entities/entities-routes/src/locales/en.json:252]:
```suggestion
    "expression_playground": {
```

nit: letter case

File: packages/core/expressions/vite.config.ts [hunk 29]
```diff
@@ -21,19 +21,19 @@ const config = mergeConfig(sharedViteConfig, defineConfig({
...
   plugins: [
     wasm(),
     topLevelAwait({
       promiseExportName: 'asyncInit',
     }),
-    // We don't need this plugin to bundle the library. Only for sandbox previews.
-    // See: https://github.com/vdesjs/vite-plugin-monaco-editor/issues/21
-    ...process.env.USE_SANDBOX
+    // This plugin is only used in the sandbox & testing environment
+    // It generates extra files in dist folder whitch are not need in library build
```
Comment (sumimakito) [inline packages/core/expressions/vite.config.ts:33]:
```suggestion
    // It generates extra files in dist folder which are not need in library build
```

**Comment by sumimakito** [comment]
Tested functionality in Kong Manager and Konnect Gateway Manager. LGTM

File: packages/entities/entities-routes/src/locales/en.json (hunk 28 above)
Comment (2eha0) [reply packages/entities/entities-routes/src/locales/en.json]: fixed

File: packages/core/expressions/vite.config.ts (hunk 29 above)
Comment (2eha0) [reply packages/core/expressions/vite.config.ts]: fixed

File: packages/core/expressions/README.md (hunk 27 above)
Comment (2eha0) [reply packages/core/expressions/README.md:63]: added

**Review by sumimakito — APPROVED** [review]

**Review by Leopoldthecoder — APPROVED** [review]
